### PR TITLE
[SPARK-6629] cancelJobGroup() may not work for jobs whose job groups are inherited from parent threads

### PR DESCRIPTION
**Pull request by JoshRosen**
When a job is submitted with a job group and that job group is inherited from a parent thread, there are multiple bugs that may prevent this job from being cancelable via `SparkContext.cancelJobGroup()`:
- When filtering jobs based on their job group properties, DAGScheduler calls `get()` instead of `getProperty()`, which does not respect inheritance, so it will skip over jobs whose job group properties were inherited.
- `Properties` objects are mutable, but we do not make defensive copies / snapshots, so modifications of the parent thread's job group will cause running jobs' groups to change; this also breaks cancelation.

Both of these issues are easy to fix: use `getProperty()` and perform defensive copying.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -493,7 +495,8 @@ class DAGScheduler(
     val func2 = func.asInstanceOf[(TaskContext, Iterator[_]) => _]
     val waiter = new JobWaiter(this, jobId, partitions.size, resultHandler)
     eventProcessLoop.post(JobSubmitted(
-      jobId, rdd, func2, partitions.toArray, allowLocal, callSite, waiter, properties))
+      jobId, rdd, func2, partitions.toArray, allowLocal, callSite, waiter,
+      SerializationUtils.clone(properties)))
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: We need to use `SerializationUtils.clone()` in order to perform a deep-clone which also captures properties that were inherited from parents; see 5a5526164bdf9ecf1306d4570e816eb4df5cfd2b for more discussion of this issue.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 2]
```diff
@@ -675,7 +679,7 @@ class DAGScheduler(
     // Cancel all jobs belonging to this job group.
     // First finds all active jobs with this group id, and then kill stages for them.
     val activeInGroup = activeJobs.filter(activeJob =>
-      groupId == activeJob.properties.get(SparkContext.SPARK_JOB_GROUP_ID))
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: This bug has been around since 2013!

**Comment by JoshRosen** [comment]
On further reflection, this might not be a use-case that we want to support.  I don't think we make any user-facing promises about how / whether these properties are inherited across threads and there are reasons why we may want to discourage this implicit inheritance in favor of explicitly setting the job group in each thread that submits jobs; see https://issues.apache.org/jira/browse/SPARK-4514 for a related discussion.

**Comment by JoshRosen** [comment]
@jerryshao, since you added unit tests for `localProperty` inheritance in ffa5f8e, do you know of any actual use-cases that motivate this inheritance?  As far as I can tell, we don't seem to rely on it internally inside of Spark and it's not documented in a user-facing way.  Users can't be relying on it for job group cancellation, since this patch demonstrates that this has always been broken, and it seems unlikely that anyone would choose to use this for cross-thread communication: in order to read and write these properties, you need to have a reference to the SparkContext instance, at which point you might as well just pass the information directly instead of implicitly inheriting it across threads.

If there's no actual use for this inheritance, it might be nice to remove it.

**Comment by jerryshao** [comment]
Hi @JoshRosen , sorry for late response, looks like I missed this notification. The reason why here changes to `InheritableThreadLocal` is that: we have such a scenario, we enabled FairScheduler in Spark with two groups, one group is running streaming program, and another group is running query, they separated in two threads and shared with same SparkContext, since streaming job is running in a threadpool when submitted to Spark DAGScheduler, so if we do not use `InheritableThreadLocal`, we cannot get the right value in child thread, so I changed to this way.

**Comment by jerryshao** [comment]
Hi @JoshRosen , I think this is the original PR (https://github.com/mesos/spark/pull/937). I'm not sure is this issue still existed, it is a quite old code.

**Comment by zsxwing** [comment]
> use getProperty() and perform defensive copying.

+1. It's possible that `SparkListener` will see some wrong key-values in `SparkListenerJobStart.properties` without the defensive copying.

**Comment by zsxwing** [comment]
@JoshRosen, will this PR be merged? I'm using `localProperties` in #5473. If we don't copy `localProperties`, `SparkListenerJobStart.properties` will just point to `localProperties`. It will be used in `SparkListener`, which runs in another thread. If `localProperties` is changed before a listener uses `SparkListenerJobStart.properties`, the listener will see the modification rather than the properties when Job starts.

**Comment by JoshRosen** [comment]
@zsxwing I haven't merged this yet because I had some doubts about how inheritance of local properties should work, especially when threads are re-used.  Let me revisit this now, though, since I think the changes here seem to make the behavior more consistent.  I'll also take a closer look at your use-case in #5473 to try to understand how it would be affected by this change and whether it's prone to any of the weird thread re-use corner cases that can crop up with localProperties.

**Comment by JoshRosen** [comment]
/cc @andrewor14, what do you think about this fix?  The inheritance + cancellation tests here address an incredibly obscure corner-case that most users might not hit or expect to work (since I don't think that we ever explicitly state that job groups will be inherited across threads).  The defensive copy of Properties seems like a good change by itself, though.

I'd like to try to get some closure on this issue by either deciding that we're going to support the inheritance feature as it was originally described or by removing this feature and instructing users to manually set job groups in child threads; at this point, I think that either option could work since the current behavior is broken.  We could also decide to just leave the inheritance alone and only do the copying change.

**Comment by zsxwing** [comment]
@JoshRosen is it ready to go? At least, #5473 needs the defensive copy of Properties to avoid breaking tests. Thank you.

**Comment by tdas** [comment]
Merging this. The defensive copying is blocker for #5473 .

**Pull request closed without merging**